lib/ansible/modules/windows/win_psmodule.ps1

```diff
@@ -90,97 +348,123 @@ Function Remove-Repository{
+# Check PowerShell version, fail if < 5.0 and required modules are not installed
+$PsVersion = $PSVersionTable.PSVersion
+if ($PsVersion.Major -lt 5 ) {
+    $PrereqModulesInstalled = Install-PrereqModule -TestInstallationOnly
+    if ( -not $PrereqModulesInstalled ) {
+        $ErrorMessage = "Modules PowerShellGet and PackageManagement in versions 1.6.0 and 1.1.7 respectively have to be installed before using the win_psmodule."
         Fail-Json $result $ErrorMessage
-      }
     }
 }
 
-Function Remove-PsModule {
-    param(
-      [Parameter(Mandatory=$true)]
-      [string]$Name,
-      [bool]$CheckMode
-    )
-    # If module is present, unistalls it.
-    if (Get-Module -Listavailable|?{$_.name -eq $Name}){
-      try{
-        Uninstall-Module -Name $Name -Confirm:$false -Force -ErrorAction Stop -WhatIf:$CheckMode | out-null
-        $result.output = "Module $($Name) removed"
-        $result.changed = $true
-      }
-      catch{
-        $ErrorMessage = "Problems removing $($Name) module: $($_.Exception.Message)"
+if ( $required_version -and ( $minimum_version -or $maximum_version ) ) {
+       $ErrorMessage = "Parameters required_version and minimum/maximum_version are mutually exclusive."
+       Fail-Json $result $ErrorMessage
+}
+
+if ( $allow_prerelease -and ( $minimum_version -or $maximum_version ) ) {
+    $ErrorMessage = "Parameters minimum_version, maximum_version can't be used with the parameter allow_prerelease."
+    Fail-Json $result $ErrorMessage
+}
+
+if ( $allow_prerelease -and $state -eq "absent" ) {
+    $ErrorMessage = "The parameter allow_prerelease can't be used with state set to 'absent'."
+    Fail-Json $result $ErrorMessage
+}
+
+if ( ($state -eq "latest") -and
+    ( $required_version -or $minimum_version -or $maximum_version ) ) {
+        $ErrorMessage = "When the parameter state is equal 'latest' you can use any of required_version, minimum_version, maximum_version."
         Fail-Json $result $ErrorMessage
-      }
+}
 
+if ( $repo -and (-not $url) ) {
+    $RepositoryExists = Get-PSRepository -Name $repo -ErrorAction Ignore
+    if ( $null -eq $RepositoryExists) {
+        $ErrorMessage = "The repository $repo doesn't exist."
+        Fail-Json $result $ErrorMessage
     }
-    else{
-      $result.output = "Module $($Name) not present"
-    }
+
 }
 
-# Check powershell version, fail if < 5.0
-$PsVersion = $PSVersionTable.PSVersion
-if ($PsVersion.Major -lt 5){
-  $ErrorMessage = "Powershell 5.0 or higher is needed"
-  Fail-Json $result $ErrorMessage
+if ( ($allow_clobber -or $allow_prerelease -or $skip_publisher_check -or
+    $required_version -or $minimum_version -or $maximum_version) ) {
+    # Update the PowerShellGet and PackageManagement modules.
+    # It's required to support AllowClobber, AllowPrerelease parameters.
+    Install-PrereqModule -CheckMode $check_mode
+}
+
+if ( $PSVersionTable.PSVersion.Major -lt 5 ) {
```

Comment:
Is there a reason why this is in an if statement, we may as well just import the module for all scenarios.

Reply:
This if statement should be replaced by simply calling or testing availability of a PSGet function. We have an environment where PS4 is not allowed to be upgraded to PS5, the separate PackageManagement on PS4 however is installed. The original module fails on the if in this case. Patched the if statement as mentioned and it works like a charm.

Reply:
@Finzzownt,
I'm not sure if I understand you. But before you clarify that please check the current code.

Reply:
The test for PowerShell Major 5 and up breaks the use of this module on Windows 2012R2 with PowerShell 4 and PackageManagement MSI installed (see https://docs.microsoft.com/bs-latn-ba/powershell/gallery/installing-psget, under _Get PowerShellGet module for PowerShell versions 3.0 and 4.0_). The MSI adds PSGet functionality to older PowerShell versions.
I would think that testing for the presence of a common PSGet cmdlet, like Install-Module would make the module work on such hosts also:
```
PS> $PSVersionTable.PSVersion.Major; if (Get-Command Install-Module -errorAction SilentlyContinue) {"yes"} else {"no"}
4
no
PS> .\Downloads\PackageManagement_x64.msi
PS> $PSVersionTable.PSVersion.Major; if (Get-Command Install-Module -errorAction SilentlyContinue) {"yes"} else {"no"}
4
yes
```

Reply:
The quote from the file win_psmodule.py included in the PR.

```
notes:
  - PowerShell modules needed
      - PowerShellGet >= 1.6.0
      - PackageManagement >= 1.1.7
  - PowerShell package provider needed
      - NuGet >= 2.8.5.201
  - On PowerShell 5.x required modules and a package provider will be updated under the first run of the win_psmodule module.
  - On PowerShell 3.x and 4.x you have to install them before using the win_psmodule.
```

Reply:
You know that feeling, like if only that Mariachi Band would come in right now instruments blazing?
Great, was looking cross eyed (was 05:00 in the morning), sorry.